Add floating base velocity limits to IK

Soontosh opened the pull request:
Added `FloatingBaseVelocityLimit` to address issue #126. I basically introduce a way to cap how fast the floating base of a robot can move. I added a limit object that converts the linear and angular velocity bounds into QP inequalities by reading the base frame's Jacobian, so that the constraints act directly on the floating base degrees of freedom.

Each `Configuration` exposes `floating_base_limit` and the IK solver automatically includes it alongside existing configuration.

Added a unit test at `test_floating_Base_velocity_limit.py`. All the checks passed (see below logs):
```
python -m unittest tests.test_floating_base_velocity_limit
.....
----------------------------------------------------------------------
Ran 5 tests in 16.586s

OK
```

stephane-caron commented on pink/limits/floating_base_velocity_limit.py:
```diff
@@ -0,0 +1,156 @@
+            rows.extend([row, -row])
+            bounds.extend([dt * bound, dt * bound])
+
+        for axis, bound in enumerate(self.angular_max):
```
We could combine these two loops into a single one.

```py
  for bound_vector, offset in [(self.linear_max, 0), (self.angular_max, 3)]:
      for axis, bound in enumerate(bound_vector):
          if not np.isfinite(bound):
              continue
          ...
```

Better yet, we could vectorize this entirely to avoid loops:

- Rather than `self.linear_max` and `self.angular_max`, we could store `self.twist_max = [linear_max, angular_max]`
- Instead of processing row by row, start from `J = jacobian`, set `J[:, :self.root_idx_v] = 0.0` and similarly `J[:, after_start:] = 0.0`
- G becomes a vstack of `[J, -J]`
- h becomes an hstack of `[self.twist_max * dt, -self.twist_max * dt]`

There would be no loops then and the code will run faster. What do you think?

stephane-caron reviewed (commented):
Thank you for this iteration :+1: There is one bug in the example, and we could make your implementation faster by working on Jacobian slices directly. Happy to discuss!

Soontosh commented:
Thank you so much for the detailed review @stephane-caron ! I apologize for the current mistakes and I will work to fix them. 

FYI I will be travelling over the next couple of weeks so it may take a while for me to respond to all your comments and to push an updated version

Soontosh commented on tests/test_floating_base_velocity_limit.py:
```diff
@@ -0,0 +1,168 @@
+            expected_rows.extend([row, -row])
+            expected_bounds.extend([self.dt * max_velocity] * 2)
+
+        expected_G = np.vstack(expected_rows)
```
I think I get what you mean, will do!

Soontosh replied:
Fixed! Pushed some changes to the test in commit [bd9fb1b](https://github.com/Soontosh/pink/commit/bd9fb1bb580f403c71ae137dddb30111f6ec5ed9)

stephane-caron commented:
> FYI I will be travelling over the next couple of weeks so it may take a while for me to respond to all your comments and to push an updated version

No worries, please work at the pace that fits your schedule. Your contributions are greatly appreciated by the community :+1: Safe travels!

Soontosh commented on pink/limits/floating_base_velocity_limit.py:
```diff
@@ -0,0 +1,156 @@
+
+        if self.root_joint_id is not None and self.frame_id is None:
+            raise ValueError(
+                f"Frame '{base_frame}' does not exist in the model."
```
Your approach would make the limit easier to use and safer on the user's side of things. I made minor changes to the codebase and modified the tests accordingly. Please see commit [d64e200](https://github.com/Soontosh/pink/commit/d64e200541026c2b04dbfe6f767b63b1c3e243e2)

Soontosh commented:
Hello @stephane-caron.

I've addressed all your concerns with 11 new commits. Please see my above comments for more context on each change. Below is a list of the changes:
- Add credit for creating the feature
- Remove unnecessary files
- Remove the sparseness feature from this PR
- Fail fast on missing and incorrect base frames
- Auto detect the root frame when none is provided
- Simplify floating base inequality construction to eliminate the duplicate loops and speed up code
- Update the planar test to check behavioral clipping
- Fix the test to ignore infinite bounds when building inside twists
- Update README accordingly

There are some simple formatting issues with ruff that I will be fixing this week. Other than that, this PR should be ready for merging.

If you have any additional questions or concerns, please let me know.

Soontosh commented:
All ruff checks now pass!
